File: src/main/java/pl/edu/pw/ee/catering_backend/orders/domain/OrdersService.java
package pl.edu.pw.ee.catering_backend.orders.domain;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.edu.pw.ee.catering_backend.infrastructure.db.MealDb;
import pl.edu.pw.ee.catering_backend.infrastructure.db.OrderDb;
import pl.edu.pw.ee.catering_backend.infrastructure.db.UserDb;
import pl.edu.pw.ee.catering_backend.infrastructure.db.repositories.MealRepository;
import pl.edu.pw.ee.catering_backend.infrastructure.db.repositories.OrderRepository;
import pl.edu.pw.ee.catering_backend.infrastructure.db.repositories.UserRepository;
import pl.edu.pw.ee.catering_backend.offers.comms.MealMapper;
import pl.edu.pw.ee.catering_backend.offers.domain.Meal;
import pl.edu.pw.ee.catering_backend.orders.comms.OrderMapper;
import pl.edu.pw.ee.catering_backend.orders.comms.dtos.AddOrderDTO;
import pl.edu.pw.ee.catering_backend.orders.comms.dtos.OrderDto;
import pl.edu.pw.ee.catering_backend.orders.infrastructure.OrderStatus;
import pl.edu.pw.ee.catering_backend.user.infrastructure.UserDbMapper;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class OrdersService implements IOrdersService {

    private final IOrdersPersistenceService ordersPersistenceService;
    private final MealRepository mealRepository;
    private final OrderRepository orderRepository;
    private final UserRepository userRepository;
    private final UserDbMapper userDbMapper;
    private final OrderMapper orderMapper;
    private final MealMapper mealMapper;

    @Transactional
    @Override
    public Order addOrder(
            String userLogin,
            AddOrderDTO addOrderDTO
    ) {
        Order order = orderMapper.mapDtoToDomainModel(addOrderDTO);
        List<MealDb> dbMeals = mealRepository.findAllById(addOrderDTO.getMealIds());
        List<Meal> mappedMeals = dbMeals.stream()
                .map(mealMapper::mapToDomain)
                .toList();
        LocalDateTime orderCreationTime = LocalDateTime.now();
        UserDb userDb = userRepository.findByLogin(userLogin).orElseThrow(() -> new NoSuchElementException("User with login " + userLogin + " not found"));
        BigDecimal totalPrice = mappedMeals.stream()
                .map(Meal::getPrice)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        if (mappedMeals.size() != addOrderDTO.getMealIds().size()) {
            List<UUID> notFoundMealIds = addOrderDTO.getMealIds().stream()
                    .filter(mealId -> dbMeals.stream().noneMatch(mealDb -> mealDb.getId().equals(mealId)))
                    .toList();
            throw new IllegalArgumentException("Some meals were not found in the database: " + notFoundMealIds);
        }

        order.setMeals(mappedMeals);
        order.setOrderCreationTime(orderCreationTime);
        order.setClient(userDbMapper.toDomain(userDb));
        order.setStatus(OrderStatus.UNPAID);
        order.setTotalPrice(totalPrice);

        order.validateForUpdate();

        Order savedOrder = ordersPersistenceService.save(order, dbMeals);

        return savedOrder;
    }

    @Override
    public List<OrderDto> getOrders(String userLogin) {
        UserDb userDb = userRepository.findByLogin(userLogin).orElseThrow(() -> new NoSuchElementException("User with login " + userLogin + " not found"));
        List<OrderDb> orders = orderRepository.findAll().stream()
                .filter(orderDb -> orderDb.getClient().equals(userDb))
                .toList();

        List<OrderDto> orderDtos = orders.stream()
                .map(orderMapper::mapDbToDomainModel)
                .map(orderMapper::mapToOrderDto)
                .toList();

        return orderDtos;
    }

    // TODO fix mapping
    @Override
    public OrderDto getOrderById(UUID id) {
        OrderDb orderDb = orderRepository.findById(id).orElseThrow(() -> new NoSuchElementException("Order with id " + id + " not found"));
        Order order = orderMapper.mapDbToDomainModel(orderDb);
        return orderMapper.mapToOrderDto(order);

    }

}

File: src/main/java/pl/edu/pw/ee/catering_backend/orders/data/OrdersService.kt
package pl.edu.pw.ee.catering_backend.orders.data

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import pl.edu.pw.ee.catering_backend.infrastructure.db.Client
import pl.edu.pw.ee.catering_backend.infrastructure.db.Meal
import pl.edu.pw.ee.catering_backend.infrastructure.db.Order
import pl.edu.pw.ee.catering_backend.infrastructure.db.repositories.MealRepository
import pl.edu.pw.ee.catering_backend.infrastructure.db.repositories.OrderRepository
import java.util.*

interface OrdersService {
    fun deleteOrder(
        orderId: Long,
    ): Result<Unit>

    fun updateOrder(
        orderId: Long,
        clientLogin: String,
        mealsIds: List<String>,
        deliveryAddress: String,
        deliveryMethod: String,
        status: String,
    ): Result<Unit>

    fun getAllOrders(
        clientLogin: String,
    ): Result<List<Order>>

    fun createOrder(
        client: Client? = null,
        mealsIds: List<Meal>,
        deliveryAddress: String,
        deliveryMethod: String,
        status: String
    ): Result<String>
}

@Service
class OrdersServiceImpl(
    @Autowired private val ordersRepository: OrderRepository,
    @Autowired private val mealRepository: MealRepository,
) : OrdersService {
    override fun createOrder(
        client: Client?,
        mealsIds: List<Meal>,
        deliveryAddress: String,
        deliveryMethod: String,
        status: String,
    ): Result<String> {
        return try {
            // Save the order to the repository
            ordersRepository.insertOrder(
                deliveryAddress,
                deliveryMethod,
                mealsIds,
                status,
                null,
            )
            Result.success(UUID.randomUUID().toString())
        } catch (e: Exception) {
            Result.success(UUID.randomUUID().toString())
        }
    }

    override fun deleteOrder(orderId: Long): Result<Unit> {
        try {
            val uuid = UUID.fromString(orderId.toString())
            ordersRepository.deleteById(uuid)
            return Result.success(Unit)
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override fun updateOrder(
        orderId: Long,
        clientLogin: String,
        mealsIds: List<String>,
        deliveryAddress: String,
        deliveryMethod: String,
        status: String,
    ): Result<Unit> {
        try {
            val order = ordersRepository.findById(UUID.fromString(orderId.toString()))
            if (order.isPresent) {
                // TODO: fix java interop
//                val updatedOrder = order.get()
//                updatedOrder.client.login = clientLogin
//                updatedOrder.meals = mealRepository.findAllById(mealsIds.map { UUID.fromString(it) })
//                updatedOrder.deliveryAddress = deliveryAddress
//                updatedOrder.deliveryMethod = deliveryMethod
//                updatedOrder.status = status
//                ordersRepository.save(updatedOrder)
                return Result.success(Unit)
            } else {
                return Result.failure(Exception("Order not found"))
            }
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override fun getAllOrders(clientLogin: String): Result<List<Order>> {
        return try {
            Result.success(ordersRepository.findByClientLogin(clientLogin))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

}